File: crypto/vss/feldman_vss.go
// Copyright © 2019 Binance
//
// This file is part of Binance. The full Binance copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

// Feldman VSS, based on Paul Feldman, 1987., A practical scheme for non-interactive verifiable secret sharing.
// In Foundations of Computer Science, 1987., 28th Annual Symposium on. IEEE, 427–43
//

package vss

import (
	"crypto/elliptic"
	"errors"
	"fmt"
	"io"
	"math/big"

	"github.com/bnb-chain/tss-lib/v4/common"
	"github.com/bnb-chain/tss-lib/v4/crypto"
	"github.com/bnb-chain/tss-lib/v4/tss"
)

type (
	Share struct {
		Threshold int
		ID,       // xi
		Share *big.Int // Sigma i
	}

	Vs []*crypto.ECPoint // v0..vt

	Shares []*Share
)

var (
	ErrNumSharesBelowThreshold = fmt.Errorf("not enough shares to satisfy the threshold")

	zero = big.NewInt(0)
	one  = big.NewInt(1)
)

// Check share ids of Shamir's Secret Sharing, return error if duplicate or 0 value found
func CheckIndexes(ec elliptic.Curve, indexes []*big.Int) ([]*big.Int, error) {
	visited := make(map[string]struct{})
	for _, v := range indexes {
		vMod := new(big.Int).Mod(v, ec.Params().N)
		if vMod.Cmp(zero) == 0 {
			return nil, errors.New("party index should not be 0")
		}
		vModStr := vMod.String()
		if _, ok := visited[vModStr]; ok {
			return nil, fmt.Errorf("duplicate indexes %s", vModStr)
		}
		visited[vModStr] = struct{}{}
	}
	return indexes, nil
}

// Returns a new array of secret shares created by Shamir's Secret Sharing Algorithm,
// requiring a minimum number of shares to recreate, of length shares, from the input secret
func Create(ec elliptic.Curve, threshold int, secret *big.Int, indexes []*big.Int, rand io.Reader) (Vs, Shares, error) {
	if ec == nil || rand == nil {
		return nil, nil, fmt.Errorf("vss Create: ec or rand == nil")
	}
	if secret == nil || indexes == nil {
		return nil, nil, fmt.Errorf("vss secret or indexes == nil: %v %v", secret, indexes)
	}
	if threshold < 1 {
		return nil, nil, errors.New("vss threshold < 1")
	}

	ids, err := CheckIndexes(ec, indexes)
	if err != nil {
		return nil, nil, err
	}

	num := len(indexes)
	// Need at least threshold+1 distinct shares to reconstruct a
	// degree-`threshold` polynomial; the old `num < threshold` check
	// admitted num == threshold which produced an unreconstructable
	// share set.
	if num < threshold+1 {
		return nil, nil, ErrNumSharesBelowThreshold
	}

	poly := samplePolynomial(ec, threshold, secret, rand)

	v := make(Vs, len(poly))
	for i, ai := range poly {
		v[i] = crypto.ScalarBaseMult(ec, ai)
	}

	shares := make(Shares, num)
	for i := 0; i < num; i++ {
		share := evaluatePolynomial(ec, threshold, poly, ids[i])
		shares[i] = &Share{Threshold: threshold, ID: ids[i], Share: share}
	}
	return v, shares, nil
}

func (share *Share) Verify(ec elliptic.Curve, threshold int, vs Vs) bool {
	if share == nil || ec == nil || share.ID == nil || share.Share == nil ||
		share.Threshold != threshold || vs == nil || len(vs) != threshold+1 {
		return false
	}
	q := ec.Params().N
	// share.ID is used as a Lagrange interpolation x-coordinate (always
	// reduced mod q downstream), so the binding requirement is "non-zero
	// mod q"; tss test IDs are random 256-bit integers that commonly
	// exceed q on Ed25519 (where q ≈ 2^252), so enforcing a canonical
	// range here would reject honest inputs.
	idModQ := new(big.Int).Mod(share.ID, q)
	if idModQ.Sign() == 0 || share.Share.Sign() <= 0 || share.Share.Cmp(q) >= 0 {
		return false
	}
	var err error
	modQ := common.ModInt(q)
	v, t := vs[0], one // YRO : we need to have our accumulator outside of the loop
	// ValidateInSubgroup rejects nil / off-curve / identity / (on
	// composite-cofactor curves) low-order points. tss.SameCurve
	// additionally requires vs[j].curve == ec rather than silently
	// re-attaching the curve via SetCurve — direct API consumers that
	// constructed vs[j] on a different elliptic.Curve instance now get
	// rejected up-front instead of having their input mutated.
	if v == nil || !tss.SameCurve(v.Curve(), ec) || !v.ValidateInSubgroup() {
		return false
	}
	for j := 1; j <= threshold; j++ {
		if vs[j] == nil || !tss.SameCurve(vs[j].Curve(), ec) || !vs[j].ValidateInSubgroup() {
			return false
		}
		// t = k_i^j
		t = modQ.Mul(t, share.ID)
		// v = v * v_j^t
		vjt := vs[j].ScalarMult(t)
		// ScalarMult returns nil for degenerate cases (identity / off-curve
		// result). Guard explicitly so Add doesn't dereference nil.
		if vjt == nil {
			return false
		}
		v, err = v.Add(vjt)
		if err != nil {
			return false
		}
	}
	sigmaGi := crypto.ScalarBaseMult(ec, share.Share)
	return sigmaGi.Equals(v)
}

func (shares Shares) ReConstruct(ec elliptic.Curve) (secret *big.Int, err error) {
	if ec == nil {
		return nil, errors.New("vss ReConstruct: ec == nil")
	}
	if len(shares) == 0 {
		return nil, ErrNumSharesBelowThreshold
	}
	// Per-share validation up-front so the Lagrange loop below can
	// assume non-nil IDs / Shares and a consistent threshold. Without
	// these checks the loop would either nil-deref or silently mix
	// shares of different threshold polynomials.
	threshold := shares[0].Threshold
	for i, share := range shares {
		if share == nil || share.ID == nil || share.Share == nil {
			return nil, fmt.Errorf("vss ReConstruct: nil share or share field at index %d", i)
		}
		if share.Threshold != threshold {
			return nil, fmt.Errorf("vss ReConstruct: share %d has threshold %d, want %d", i, share.Threshold, threshold)
		}
	}
	if threshold+1 > len(shares) {
		return nil, ErrNumSharesBelowThreshold
	}
	modN := common.ModInt(ec.Params().N)

	// x coords
	xs := make([]*big.Int, 0, len(shares))
	for _, share := range shares {
		xs = append(xs, share.ID)
	}

	secret = zero
	for i, share := range shares {
		times := one
		for j := 0; j < len(xs); j++ {
			if j == i {
				continue
			}
			sub := modN.Sub(xs[j], share.ID)
			// Reject mod-q ID collision instead of triggering
			// ModInverse(0, q) → nil → modN.Mul nil-receiver panic.
			// Same shape as the dcb631d fix in signing/prepare.go;
			// ReConstruct isn't on the production code path but is
			// exported, so external callers shouldn't be able to crash
			// the library by submitting shares with k vs k+q IDs.
			if sub.Sign() == 0 {
				return nil, fmt.Errorf("vss ReConstruct: shares %d and %d have IDs that collide mod q", j, i)
			}
			subInv := modN.ModInverse(sub)
			div := modN.Mul(xs[j], subInv)
			times = modN.Mul(times, div)
		}

		fTimes := modN.Mul(share.Share, times)
		secret = modN.Add(secret, fTimes)
	}

	return secret, nil
}

func samplePolynomial(ec elliptic.Curve, threshold int, secret *big.Int, rand io.Reader) []*big.Int {
	q := ec.Params().N
	v := make([]*big.Int, threshold+1)
	v[0] = secret
	for i := 1; i <= threshold; i++ {
		ai := common.GetRandomPositiveInt(rand, q)
		v[i] = ai
	}
	return v
}

// Evauluates a polynomial with coefficients such that:
// evaluatePolynomial([a, b, c, d], x):
//
//	returns a + bx + cx^2 + dx^3
func evaluatePolynomial(ec elliptic.Curve, threshold int, v []*big.Int, id *big.Int) (result *big.Int) {
	q := ec.Params().N
	modQ := common.ModInt(q)
	result = new(big.Int).Set(v[0])
	X := big.NewInt(int64(1))
	for i := 1; i <= threshold; i++ {
		ai := v[i]
		X = modQ.Mul(X, id)
		aiXi := new(big.Int).Mul(ai, X)
		result = modQ.Add(result, aiXi)
	}
	return
}

File: tss-lib-rust/src/crypto/vss/feldman_vss.rs
// Copyright © 2019 Binance
//
// This file is part of Binance. The full Binance copyright notice, including
// terms governing use, modification, and redistribution, is contained in the
// file LICENSE at the root of the source code distribution tree.

// Feldman VSS, based on Paul Feldman, 1987., A practical scheme for non-interactive verifiable secret sharing.
// In Foundations of Computer Science, 1987., 28th Annual Symposium on. IEEE, 427–43

// Translation of tss-lib-go/crypto/vss/feldman_vss.go

use crate::{
    common::{
        int::ModInt,
        random::get_random_positive_int,
    },
    crypto::ecpoint::{ECPoint, PointError}, // Assuming generic ECPoint
    tss::Curve, // Assuming trait for curve operations & params
};

use elliptic_curve::CurveArithmetic;
use elliptic_curve::scalar::Scalar;
use num_bigint_dig::{{BigInt, Sign}};
use num_traits::{{Zero, One}};
use rand::{{CryptoRng, RngCore}};
use serde::{{Deserialize, Serialize}};
use std::collections::HashSet;
use thiserror::Error;
use log::warn;


#[derive(Error, Debug, Clone, PartialEq, Eq)]
pub enum VssError {
    #[error("not enough shares to satisfy the threshold (required: {0}, have: {1})")]
    NumSharesBelowThreshold(usize, usize),
    #[error("invalid parameters: {0}")]
    InvalidParameters(String),
    #[error("duplicate party index found: {0}")]
    DuplicateIndex(String),
    #[error("party index is zero, which is not allowed")]
    IndexIsZero,
    #[error("share verification failed for party {id}")]
    ShareVerificationError { id: BigInt },
    #[error("reconstruction failed: {0}")]
    ReconstructionError(String),
    #[error("point operation failed: {0}")]
    PointError(String),
}

impl From<PointError> for VssError {
    fn from(err: PointError) -> Self {
        VssError::PointError(err.to_string())
    }
}

/// Represents a VSS Share σᵢ for a party Pᵢ.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct Share {
    pub threshold: usize,
    #[serde(with = "crate::serde_support::bigint_bytes")]
    pub id: BigInt, // Corresponds to x_i
    #[serde(with = "crate::serde_support::bigint_bytes")]
    pub share: BigInt, // Corresponds to σ_i = f(x_i)
}

/// Represents the public verification vector V = [v₀, v₁, ..., vₜ] where vᵢ = g^aᵢ.
#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct VerificationVector<C: Curve + CurveArithmetic> {
     // Need to handle ECPoint serialization carefully
     #[serde(bound(serialize = "ECPoint<C>: Serialize", deserialize = "ECPoint<C>: Deserialize<'de>"))]
    pub vector: Vec<ECPoint<C>>,
}

/// Convenience type for a slice of Shares.
pub type Shares<'a> = &'a [Share];

/// Checks share IDs (indexes) for duplicates or zero values modulo the curve order `q`.
/// Returns the original indexes if valid.
pub fn check_indexes<
    C: Curve + CurveArithmetic
>(q: &BigInt, indexes: &[BigInt]) -> Result<(), VssError> {
    if indexes.is_empty() {
         // Or should this be allowed?
         return Err(VssError::InvalidParameters("Indexes slice cannot be empty".to_string()));
     }
    let mut visited = HashSet::new();
    let mod_q = ModInt::new(q.clone());

    for v in indexes {
        // Reduce index modulo q. Does ID need to be < q?
        // Original Go code reduces, let's follow that for now.
        let v_mod = mod_q.add(v, &BigInt::zero()); // Ensure positive result < q

        if v_mod.is_zero() {
            return Err(VssError::IndexIsZero);
        }
        let v_mod_str = v_mod.to_string();
        if !visited.insert(v_mod_str) {
            return Err(VssError::DuplicateIndex(v_mod.to_string()));
        }
    }
    Ok(())
}

/// Creates a new Feldman Verifiable Secret Sharing scheme.
/// Generates shares of the `secret` for parties identified by `indexes`,
/// with a given `threshold`.
/// Returns the verification vector `V` and the list of shares `σᵢ`.
pub fn create<
    C: Curve + CurveArithmetic,
    R: CryptoRng + RngCore,
>(
    q: &BigInt, // Curve order
    threshold: usize,
    secret: &BigInt,
    indexes: &[BigInt],
    rng: &mut R,
) -> Result<(VerificationVector<C>, Vec<Share>), VssError>
where
     // Bounds needed for ECPoint operations
     ECPoint<C>: Clone + PartialEq + Serialize + for<'de> Deserialize<'de>,
     // Assuming ECPoint::scalar_base_mult exists and takes BigInt
     // Assuming ECPoint implements Add and ScalarMult traits or methods
{
    if threshold < 1 {
        return Err(VssError::InvalidParameters("Threshold cannot be less than 1".to_string()));
    }
    if indexes.len() < threshold + 1 {
        return Err(VssError::NumSharesBelowThreshold(threshold + 1, indexes.len()));
    }
    check_indexes::<C>(q, indexes)?;

    // 1. Sample polynomial f(z) = a₀ + a₁z + ... + aₜzᵗ where a₀ = secret
    let poly = sample_polynomial(q, threshold, secret, rng);

    // 2. Compute verification vector V = [g^a₀, g^a₁, ..., g^aₜ]
    let v_vec: Vec<ECPoint<C>> = poly.iter()
        .map(|a_i| ECPoint::<C>::scalar_base_mult(a_i))
        .collect();
    let verification_vector = VerificationVector { vector: v_vec };

    // 3. Compute shares σᵢ = f(idᵢ) for each party i
    let shares_vec: Vec<Share> = indexes.iter()
        .map(|id| {
            let share_val = evaluate_polynomial(q, threshold, &poly, id);
            Share {
                threshold,
                id: id.clone(), // Use original ID provided
                share: share_val,
            }
        })
        .collect();

    Ok((verification_vector, shares_vec))
}

impl Share {
    /// Verifies a share `σᵢ` against the public verification vector `V`.
    /// Checks if g^σᵢ = Π (vⱼ)^(idᵢ^j) for j = 0 to t.
    pub fn verify<
        C: Curve + CurveArithmetic
    >(
        &self,
        q: &BigInt, // Curve order
        verification_vector: &VerificationVector<C>,
    ) -> bool
    where
         // Bounds needed for ECPoint operations
         ECPoint<C>: Clone + PartialEq + Serialize + for<'de> Deserialize<'de>,
         // Assuming ECPoint implements Add and ScalarMult traits or methods
    {
        if self.threshold + 1 != verification_vector.vector.len() {
             warn!("Share verify failed: threshold mismatch (share={}, vv={})", self.threshold, verification_vector.vector.len()-1);
            return false;
        }

        let vs = &verification_vector.vector;
        let mod_q = ModInt::new(q.clone());

        // Calculate the right side of the equation: Π (vⱼ)^(idᵢ^j) mod N
        // rhs = v₀ * v₁^id * v₂^(id²) * ... * vₜ^(idᵗ)

        let mut rhs = vs[0].clone(); // Initialize with v₀ = g^a₀
        let mut id_power_j = BigInt::one();

        for j in 1..=self.threshold {
            // id_power_j = id^j mod q
            id_power_j = mod_q.mul(&id_power_j, &self.id);

            // point_j = v_j ^ (id^j)
            let point_j = vs[j].scalar_mul(&id_power_j);

            // rhs = rhs + point_j (point addition)
            match rhs.add(&point_j) {
                 Ok(sum) => rhs = sum,
                 Err(_) => {
                     warn!("Share verify failed: point addition error during RHS calculation");
                     return false; // Error during point addition
                 }
             }
        }

        // Calculate the left side: g^σᵢ
        let lhs = ECPoint::<C>::scalar_base_mult(&self.share);

        // Compare lhs == rhs
        lhs == rhs
    }
}

/// Reconstructs the secret from a sufficient number of shares using Lagrange interpolation.
pub fn reconstruct_secret<
    C: Curve + CurveArithmetic
>(
    q: &BigInt, // Curve order
    shares: Shares,
) -> Result<BigInt, VssError> {
    if shares.is_empty() {
        return Err(VssError::ReconstructionError("Cannot reconstruct secret from empty shares".to_string()));
    }

    let threshold = shares[0].threshold;
    if shares.len() <= threshold {
        return Err(VssError::NumSharesBelowThreshold(threshold + 1, shares.len()));
    }

    // Use only the first t+1 shares for reconstruction
    let effective_shares = &shares[0..=threshold];
    let mod_q = ModInt::new(q.clone());

    let mut secret = BigInt::zero();

    for i in 0..effective_shares.len() {
        let share_i = &effective_shares[i];
        let id_i = &share_i.id;

        // Calculate Lagrange basis polynomial lᵢ(0)
        let mut lagrange_basis = BigInt::one();
        for j in 0..effective_shares.len() {
            if i == j {
                continue;
            }
            let id_j = &effective_shares[j].id;

            // term = id_j / (id_j - id_i) mod q
            let denominator = mod_q.sub(id_j, id_i);
            if denominator.is_zero() {
                 // This should not happen if check_indexes passed and IDs are distinct mod q
                 return Err(VssError::ReconstructionError(format!(
                     "Lagrange denominator is zero for i={}, j={} (id_i={}, id_j={})",
                     i, j, id_i, id_j
                 )));
             }
            let denominator_inv = mod_q.mod_inverse(&denominator)
                .ok_or_else(|| VssError::ReconstructionError(format!(
                     "Modular inverse failed for denominator (id_j - id_i) = {} mod {} for i={}, j={}",
                     denominator, q, i, j
                 )))?;

            let term = mod_q.mul(id_j, &denominator_inv);
            lagrange_basis = mod_q.mul(&lagrange_basis, &term);
        }

        // Add term: shareᵢ * lᵢ(0) mod q
        let term_i = mod_q.mul(&share_i.share, &lagrange_basis);
        secret = mod_q.add(&secret, &term_i);
    }

    Ok(secret)
}

// --- Private Helper Functions ---

/// Samples a random polynomial f(z) = a₀ + a₁z + ... + aₜzᵗ of degree `threshold`,
/// where a₀ = `secret` and other coefficients a₁, ..., aₜ are random values in Zq.
fn sample_polynomial<
    R: CryptoRng + RngCore
>(
    q: &BigInt, // Curve order
    threshold: usize,
    secret: &BigInt,
    rng: &mut R,
) -> Vec<BigInt> {
    let mut poly = Vec::with_capacity(threshold + 1);
    // a₀ = secret (reduced mod q? VSS normally operates in the field Zq)
    let mod_q = ModInt::new(q.clone());
    poly.push(mod_q.add(secret, &BigInt::zero())); // Ensure a0 is in Zq

    // a₁, ..., aₜ = random in Zq
    for _ in 1..=threshold {
        // Ensure coefficient is less than q
        let ai = get_random_positive_int(rng, q).unwrap_or_else(BigInt::zero);
        poly.push(ai);
    }
    poly
}

/// Evaluates the polynomial `poly` at point `id` modulo `q`.
/// poly = [a₀, a₁, ..., aₜ]
/// result = a₀ + a₁*id + a₂*id² + ... + aₜ*idᵗ mod q
fn evaluate_polynomial(
    q: &BigInt, // Curve order
    _threshold: usize, // Not strictly needed if poly length implies it
    poly: &[BigInt],
    id: &BigInt,
) -> BigInt {
    let mod_q = ModInt::new(q.clone());
    let mut result = poly.get(0).cloned().unwrap_or_else(BigInt::zero);
    let mut id_power_i = BigInt::one();

    for i in 1..poly.len() {
        // id_power_i = id^i mod q
        id_power_i = mod_q.mul(&id_power_i, id);

        // term = aᵢ * id^i mod q
        let term = mod_q.mul(&poly[i], &id_power_i);

        // result = (result + term) mod q
        result = mod_q.add(&result, &term);
    }
    result
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::tss::Secp256k1Curve; // Example curve
    use elliptic_curve::Field;
    use rand::thread_rng;

     // Helper to get curve order Q for K256
     fn get_k256_q() -> BigInt {
         let q_bytes = k256::Scalar::ORDER.to_be_bytes();
         BigInt::from_bytes_be(Sign::Plus, &q_bytes)
     }

    #[test]
    fn test_feldman_vss_create_verify_reconstruct() {
        let mut rng = thread_rng();
        let q = get_k256_q();

        let secret = BigInt::from(123456789012345_u64);
        let threshold = 2;
        let num_parties = 5;

        // Generate party IDs (must be > 0 and distinct mod q)
        let indexes: Vec<BigInt> = (1..=num_parties).map(BigInt::from).collect();

        // 1. Create VSS shares
        let create_result = create::<Secp256k1, _>(&q, threshold, &secret, &indexes, &mut rng);
        assert!(create_result.is_ok());
        let (verification_vector, shares_vec) = create_result.unwrap();

        assert_eq!(verification_vector.vector.len(), threshold + 1);
        assert_eq!(shares_vec.len(), num_parties as usize);

        println!("Secret: {}", secret);
        println!("Threshold: {}", threshold);
        println!("Verification Vector V[0]: {:?}", verification_vector.vector[0]); // g^a0 = g^secret
        // println!("Shares: {:?}", shares_vec);

        // 2. Verify each share
        for share in &shares_vec {
            println!("Verifying share for ID: {}", share.id);
            assert!(share.verify::<Secp256k1>(&q, &verification_vector), "Share verification failed for ID {}", share.id);
        }

        // 3. Reconstruct secret with enough shares (t+1)
        let shares_to_reconstruct = &shares_vec[0..=threshold];
        let reconstructed_secret = reconstruct_secret::<Secp256k1>(&q, shares_to_reconstruct)
            .expect("Secret reconstruction failed");

        println!("Reconstructed Secret: {}", reconstructed_secret);
        assert_eq!(secret, reconstructed_secret);

         // 4. Reconstruct with different set of t+1 shares
         let shares_to_reconstruct_alt = &shares_vec[num_parties as usize - threshold - 1..];
         assert_eq!(shares_to_reconstruct_alt.len(), threshold + 1);
         let reconstructed_secret_alt = reconstruct_secret::<Secp256k1>(&q, shares_to_reconstruct_alt)
             .expect("Secret reconstruction (alt set) failed");
         assert_eq!(secret, reconstructed_secret_alt);

        // 5. Attempt reconstruction with insufficient shares (t)
        let shares_insufficient = &shares_vec[0..threshold];
        let recon_insufficient = reconstruct_secret::<Secp256k1>(&q, shares_insufficient);
        assert!(matches!(recon_insufficient, Err(VssError::NumSharesBelowThreshold(_, _))));
    }

    #[test]
    fn test_verify_fail_tampered_share() {
        let mut rng = thread_rng();
        let q = get_k256_q();
        let secret = BigInt::from(999u64);
        let threshold = 1;
        let indexes = vec![BigInt::from(1), BigInt::from(2)];

        let (vv, mut shares_vec) = create::<Secp256k1, _>(&q, threshold, &secret, &indexes, &mut rng).unwrap();

        // Tamper with a share value
        shares_vec[0].share += BigInt::one();

        assert!(!shares_vec[0].verify::<Secp256k1>(&q, &vv), "Tampered share verified successfully");
        assert!(shares_vec[1].verify::<Secp256k1>(&q, &vv), "Untampered share failed verification");
    }

    #[test]
    fn test_verify_fail_wrong_vv() {
         let mut rng = thread_rng();
        let q = get_k256_q();
        let secret = BigInt::from(888u64);
        let threshold = 1;
        let indexes = vec![BigInt::from(1), BigInt::from(2)];

        let (vv1, shares1) = create::<Secp256k1, _>(&q, threshold, &secret, &indexes, &mut rng).unwrap();
        // Create shares/VV for a *different* secret
        let secret2 = BigInt::from(777u64);
        let (vv2, _) = create::<Secp256k1, _>(&q, threshold, &secret2, &indexes, &mut rng).unwrap();

        // Try to verify shares from secret1 against vv from secret2
        assert!(!shares1[0].verify::<Secp256k1>(&q, &vv2), "Share verified against wrong VV");
        assert!(shares1[0].verify::<Secp256k1>(&q, &vv1), "Share failed against correct VV");
    }

     #[test]
    fn test_check_indexes() {
        let q = get_k256_q();
        assert!(check_indexes::<Secp256k1>(&q, &[BigInt::one(), BigInt::two()]).is_ok());
        // Zero index
        assert!(matches!(check_indexes::<Secp256k1>(&q, &[BigInt::one(), BigInt::zero()]), Err(VssError::IndexIsZero)));
        // Duplicate index
        assert!(matches!(check_indexes::<Secp256k1>(&q, &[BigInt::one(), BigInt::two(), BigInt::one()]), Err(VssError::DuplicateIndex(_))));
         // Duplicate index (after mod q)
         let q_plus_1 = &q + BigInt::one();
         assert!(matches!(check_indexes::<Secp256k1>(&q, &[BigInt::one(), q_plus_1]), Err(VssError::DuplicateIndex(_))));
         // Empty
         assert!(matches!(check_indexes::<Secp256k1>(&q, &[]), Err(VssError::InvalidParameters(_))));

    }

    #[test]
    fn test_create_errors() {
        let mut rng = thread_rng();
        let q = get_k256_q();
        let secret = BigInt::from(1u64);
        let indexes = vec![BigInt::one(), BigInt::two()];

        // Threshold too low
        assert!(matches!(create::<Secp256k1, _>(&q, 0, &secret, &indexes, &mut rng), Err(VssError::InvalidParameters(_))));

        // Not enough indexes for threshold
        assert!(matches!(create::<Secp256k1, _>(&q, 2, &secret, &indexes, &mut rng), Err(VssError::NumSharesBelowThreshold(_, _))));

        // Invalid indexes
        let invalid_indexes = vec![BigInt::one(), BigInt::zero()];
        assert!(matches!(create::<Secp256k1, _>(&q, 1, &secret, &invalid_indexes, &mut rng), Err(VssError::IndexIsZero)));
    }
} 